src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -106,6 +112,13 @@ private Task SetRoutingAndContinue(HttpContext httpContext)
             }
             else
             {
+                // Raise an event if the route matched
+                if (_diagnosticListener.IsEnabled() && _diagnosticListener.IsEnabled(DiagnosticsRouteMatchedKey))
+                {
+                    // We're just going to send the HttpContext since it has all of the relevant information
+                    _diagnosticListener.Write(DiagnosticsRouteMatchedKey, httpContext);
```

Comment:
>  // We're just going to send the HttpContext since it has all of the relevant information

How does that actually work? Does someone need the `HttpContext` in their code to do anything useful?

Reply:
What do you mean? I think the answer is yes but I’m not sure what you’re asking.

Reply:
> How does that actually work?

The person who subscribes to the event will get the HttpContext. They can use it to log additional information, e.g. when the event is raised they can get the endpoint from the HttpContext and write its name/metadata.

Reply:
>The person who subscribes to the event will get the HttpContext. They can use it to log additional information, e.g. when the event is raised they can get the endpoint from the HttpContext and write its name/metadata.

I invented `DiagnosticSource` 😆 

What I mean is - do we now expect users to have access to the 3.0 `HttpContext` type when writing listener code? We used to pass a bunch of a few pieces of top-level data when we defined events because it's not always easy to retrieve data using reflection on the `HttpContext` if it's more complex than a property.

Has the landscape changed? Does ApplicationInsights just reference our stuff directly?

Reply:
The other point is that if we ever wanted to pass some other data here as this event, it wouldn't be possible without a breaking change.

Reply:
> I invented DiagnosticSource

TIL 😋 

Won't passing an anonymous object end up requiring reflection anyway?

Reply:
That's why I'm asking. The usage patterns have shifted a bit since we invented it.

Reply:
Looks like reflection on all the things: https://github.com/open-telemetry/opentelemetry-dotnet/blob/3538e05f0eed9df008d30e82a70bef11bd579bb5/src/OpenTelemetry.Collector.AspNetCore/Implementation/HttpInListener.cs

Reply:
> What I mean is - do we now expect users to have access to the 3.0 HttpContext type when writing listener code? We used to pass a bunch of a few pieces of top-level data when we defined events because it's not always easy to retrieve data using reflection on the HttpContext if it's more complex than a property.
Has the landscape changed? Does ApplicationInsights just reference our stuff directly?

@davidfowl made it HttpContext only in https://github.com/aspnet/AspNetCore/pull/11020 for perf and made it backward compatible by adding `HttpContext.HttpContext`

AppInsights piece for AspNetCore depends on Http Abstractions and knows about HttpContext. It uses reflection to fetch it from anonymous payload, but we'll switch to type cast for 3.0

Reply:
Yes people will have the HttpContext.

Reply:
> The other point is that if we ever wanted to pass some other data here as part of this event, it wouldn't be possible without a breaking change.

Yea, but I rather not allocate. Other values can be shuttled via the `HttpContext`.

Reply:
> That's why I'm asking. The usage patterns have shifted a bit since we invented it.

Right, it's unrealistic to mirror the entire object graph though there are versioning concerns around this in general, the http context is a pretty good exchange type to depend on. To get more interesting information the consumer will also need to look at specific endpoint metadata.